### PR TITLE
feat(insights): user token management

### PR DESCRIPTION
**Pull request by VladislavFitz**
| Q               | A        |
|-----------------|----------|
| Bug fix?        | no   |
| New feature?    | yes   |
| BC breaks?      | no       |
| Need Doc update | yes   |

## Describe your change
Implements Insights user-token management
- An anonymous user token is automatically generated on app launch using the same approach as its web counterpart.
- A custom user token can be set via static `Insights.userToken` property
- By default user token stored in memory, but can be stored persistently by changing  the `Insights.userTokenLeaseTime` value to `0` (default is 1440 mins = 1 day).

### CODE REVIEW

File: helper_dart/pubspec.lock [hunk 1]
```diff
@@ -641,4 +500,3 @@ packages:
     version: "3.1.1"
 sdks:
   dart: ">=2.17.5 <3.0.0"
-  flutter: ">=3.0.0"
```
Comment (Haroenv) [inline helper_dart/pubspec.lock]: is this expected?

File: helper_dart/lib/src/insights.dart [hunk 2]
```diff
@@ -17,44 +17,45 @@ class Insights implements EventTracker {
...
+  static set allowPersistentUserTokenStorage(bool isAllowed) {
+    _userTokenController.allowPersistentUserTokenStorage = isAllowed;
+  }
+
   /// Map storing Insights instances per application ID.
   static final Map<String, Insights> _insightsPool = <String, Insights>{};
 
   /// Entity managing the user token generation and storage
-  static final _userTokenController = UserTokenController();
+  static final _userTokenController = UserTokenStorage();
```
Comment (Haroenv) [inline helper_dart/lib/src/insights.dart:44]: shouldn't this be called _userTokenStorage now?

File: helper_dart/test/user_token_storage_test.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import 'package:algolia_helper/src/user_token_storage.dart';
+import 'package:test/test.dart';
+
+void main() {
+
+  test('check user token persistent storage', () async {
+    final userTokenController = UserTokenStorage()
```
Comment (Haroenv) [inline helper_dart/test/user_token_storage_test.dart:6]: here also an instance of the storage class is called a controller. Seems odd to me

File: helper_dart/lib/src/user_token_storage.dart [hunk 4]
```diff
@@ -1,17 +1,39 @@
-import 'package:shared_preferences/shared_preferences.dart';
+import 'package:hive/hive.dart';
```
Comment (Haroenv) [inline helper_dart/lib/src/user_token_storage.dart:1]: is hive builtin or just commonly used?

File: helper_dart/lib/src/user_token_storage.dart [hunk 5]
```diff
@@ -1,17 +1,39 @@
...
 
-class UserTokenController {
+import 'extensions.dart';
+
+class UserTokenStorage {
+  static const _boxName = 'userToken';
   static const _userTokenKey = 'insights-user-token';
   static const _expirationDateKey = 'insights-user-token-expiration-date';
 
+  String _userToken = 'anonymous-${const Uuid().v4()}';
```
Comment (Haroenv) [inline helper_dart/lib/src/user_token_storage.dart:9]: I'd call this defaultUserToken or anonymousUserToken

File: helper_dart/lib/src/insights.dart (hunk 2 above)
Comment (VladislavFitz) [reply helper_dart/lib/src/insights.dart]: Yes, it's a left-over.

File: helper_dart/pubspec.lock (hunk 1 above)
Comment (VladislavFitz) [reply helper_dart/pubspec.lock]: Yes, it was here due to `shared_preferences` dependency, and this it the reason why it was replaced by `hive`. helper_dart is a pure Dart package and we don't want it to depend on Flutter.

File: helper_dart/lib/src/user_token_storage.dart (hunk 4 above)
Comment (VladislavFitz) [reply helper_dart/lib/src/user_token_storage.dart:1]: It's [commonly used](https://pub.dev/packages/hive)  (top 1% of most used packages).

File: helper_dart/lib/src/insights.dart [hunk 6]
```diff
@@ -23,31 +17,57 @@ class Insights implements EventTracker {
...
+  }
 
-  /// Set custom user token
-  void setUserToken(String userToken) {
-    this.userToken = userToken;
+  /// Map storing Insights instances per application ID.
+  static final Map<String, Insights> _insightsPool = <String, Insights>{};
+
+  /// Entity managing the user token generation and storage
+  static final _userTokenStorage = UserTokenStorage();
```
Comment (aallam) [inline helper_dart/lib/src/insights.dart:44]: **Nitpick:**

For testing purposes, I would prefer `userTokenStorage` to be injectable as a parameter, and create a default static instance in `UserTokenStorage` to have a default/single instance.

For example:

1. Add it to `custom` constructor.
```dart
Insights.custom(this.service, this.userTokenStorage) : isEnabled = true;
```
2. Create a singleton instance:
```dart
class UserTokenStorage {
    final static Default = UserTokenStorage() // either as a field
    factory UserTokenStorage() = Default // or using factory, if we need to have only one instance all the time.
}
```
3. In `Insights` constructor add:
```dart
final insights = Insights.custom(
  AlgoliaEventServiceAdapter(applicationID, apiKey),
  UserTokenStorage.Default, // default impl or singleton
);
```
4. In tests, it is possible now to mock the storage, and test insights independently from the stocking implementation.

**Review by aallam — APPROVED** [review]
This is nice! LGTM!

File: helper_dart/lib/src/insights.dart (hunk 6 above)
Comment (VladislavFitz) [reply helper_dart/lib/src/insights.dart:44]: Great idea, I like it!

Comment (VladislavFitz) [reply helper_dart/lib/src/insights.dart]: Done 👌

File: helper_dart/lib/src/hits_searcher.dart [hunk 7]
```diff
@@ -226,6 +226,7 @@ class _HitsSearcher with DisposableMixin implements HitsSearcher {
       ..add(
         _responses.listen((value) {
           eventTracker.trackViews(
+            snapshot().indexName,
```
Comment (Haroenv) [inline helper_dart/lib/src/hits_searcher.dart:229]: named arguments may be nicer for trackViews

**Review by Haroenv — APPROVED** [review]

File: helper_dart/lib/src/hits_searcher.dart (hunk 7 above)
Comment (VladislavFitz) [reply helper_dart/lib/src/hits_searcher.dart]: Good point! Fixed ✅

**Pull request merged**